### PR TITLE
constructors in react components should accept props

### PR DESCRIPTION
**Pull request by Justicea83**
### Fixed Issues
$https://github.com/Expensify/App/issues/7821

### Tests
1. Confirm all tests still pass

### CODE REVIEW

**Comment by Justicea83** [comment]
cc @madmax330

**Comment by Justicea83** [comment]
updated and waiting for another review @marcaaron

**Review by marcaaron — CHANGES_REQUESTED** [review]

File: eslint-plugin-expensify/CONST.js
Comment (Justicea83) [reply eslint-plugin-expensify/CONST.js:13]: sure, I'll update it

**Comment by Justicea83** [comment]
> @Justicea83 What do you think about this comment [#49 (comment)](https://github.com/Expensify/eslint-config-expensify/pull/49#issuecomment-1071140100) and should we add another test case for that?

We shouldn't be able to as per the rules in the GH

**Comment by roryabraham** [comment]
@marcaaron [this example](https://github.com/Expensify/App/blob/2c7c82cbe534e931b85a737a86aeb20bbef058ac/src/styles/animation/SpinningIndicatorAnimation.js#L4-L12) does not extend the `React.Component` class, so I would think that it would be ignored by this lint rule.

But the example you wrote up w/ the `Meals` component would cause a lint error.

**Comment by marcaaron** [comment]
Ah thanks, I missed that it did not extend `React.Component` 😄  

So, is it just anything that calls `this` in a `constructor()` that `extends` something without calling `super()`? 
And in React components specifically `super()`must be called with `props`. 

Those are two separate things? I'm still confused why we need this and hope someone could explain it.

**Comment by Justicea83** [comment]
> ### Fixed Issues
> $[Expensify/App#7821](https://github.com/Expensify/App/issues/7821)
> 
> ### Tests
> 1. Confirm all tests still pass

@roryabraham what do you think?

**Comment by Justicea83** [comment]
Gentle bump @marcaaron

**Comment by marcaaron** [comment]
Thanks @Justicea83. I am still waiting for a response to [this comment](https://github.com/Expensify/eslint-config-expensify/pull/49#issuecomment-1071845845) and then re-review. 

I still see examples like [this](https://github.com/Expensify/App/blob/36ae0cf26f66b571e08f7d9363b4ee5afad372a2/src/components/GrowlNotification/index.js#L33-L35) and I'm unsure if we care about them or not. Do we need to provide `props` to that constructor? Or not? If yes, why? Still trying to figure out whether we need this at all 😄

**Comment by roryabraham** [comment]
To be honest, I don't think there's a _strong_ need for this lint issue. AFAIK the only case this rule really covers is if someone tries to access `this.props` in a class-style component without calling `super(props)` in the constructor. Incidentally, I think leaving out the constructor entirely (using the default constructor) is also valid.

Cases like [this](https://github.com/Expensify/App/blob/36ae0cf26f66b571e08f7d9363b4ee5afad372a2/src/components/GrowlNotification/index.js#L33-L35) technically are fine because the component does not access `this.props` anywhere. That said, there's no harm in including `super(props)` in the constructor, and it's probably better for consistency and in case someone adds a prop and tries to access it in the class component without adding `super(props)` in the constructor.

So overall I have no issues if we prefer to just close this, but if we _do_ want to keep it, the linter should only complain if all of the above are true:

1. A class extends `React.Component` or `Component`
1. The class has a constructor
1. The constructor does not take a props argument and call `super(props)`

All other cases should pass. Looking at the test cases @Justicea83 has included in this PR, we would just want to add a test case for a React class component that does not have a constructor at all (that should pass).

Given that @Justicea83 already put in the work and has a mostly working implementation, I personally would lean towards merging it because it will only be helpful. But I also don't think we'll sorely miss it if we decide to just close this.

**Comment by marcaaron** [comment]
>That said, there's no harm in including super(props) in the constructor, and it's probably better for consistency and in case someone adds a prop and tries to access it in the class component without adding super(props) in the constructor.

I can get down with the consistency argument. I just want to make sure there is some documented understanding about why we did this.

I personally would skip it because any incorrect usage of `this` inside a constructor is already covered by the rules of JS itself and a separate lint rule





And while there are valid cases where we could call `super()` with props passed in the constructor or without, there are only two cases where people actually called it without. So I guess that's just a long way of saying we _probably_ don't need this 😄

**Comment by Justicea83** [comment]
@marcaaron what do you suggest we do next?

**Comment by Justicea83** [comment]
gentle bump @marcaaron

**Comment by roryabraham** [comment]
I'm fine with just closing this and the linked issue too. It's not a problem we've really experienced AFAIK

**Pull request closed without merging**
